pkg/aspect/metricsManager.go

```diff
@@ -0,0 +1,227 @@
+			_ = env.Logger().Errorf("Failed to convert metric descriptor '%s' to definition with err: %s; skipping it.", d.Name, err)
+			continue
+		}
+
+		metric, found := find(params.Metrics, def.Name)
+		if !found {
+			env.Logger().Warningf("No metric found for descriptor %s, skipping it", def.Name)
+			continue
+		}
+
+		defs[i] = *def
+		metadata[def.Name] = &metricInfo{
+			metricKind: def.Kind,
+			value:      metric.Value,
+			labels:     metric.Labels,
+		}
+	}
+
+	asp, err := a.(adapter.MetricsBuilder).NewMetricsAspect(env, c.Builder.Params.(adapter.AspectConfig), defs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct metrics aspect with config '%v' and err: %s", c, err)
+	}
+	return &metricsWrapper{asp, metadata}, nil
+}
+
+func (*metricsManager) Kind() Kind                          { return MetricsKind }
+func (*metricsManager) DefaultConfig() adapter.AspectConfig { return &aconfig.MetricsParams{} }
+
+func (*metricsManager) ValidateConfig(adapter.AspectConfig) (ce *adapter.ConfigErrors) {
+	// TODO: we need to be provided the metric descriptors in addition to the metrics themselves here, so we can do type assertions.
+	// We also need some way to assert the type of the result of evaluating an expression, but we don't have any attributes or an
+	// evaluator on hand.
+
+	// TODO: verify all descriptors can be marshalled into istio structs (DefinitionFromProto)
+
+	return
+}
+
+func (w *metricsWrapper) Execute(attrs attribute.Bag, mapper expr.Evaluator) (*Output, error) {
+	result := &multierror.Error{}
+	var values []adapter.Value
+
+	for name, md := range w.metadata {
+		metricValue, err := mapper.Eval(md.value, attrs)
+		if err != nil {
+			result = multierror.Append(result, fmt.Errorf("failed to eval metric value for metric '%s' with err: %s", name, err))
+			continue
+		}
+		labels, err := evalAll(md.labels, attrs, mapper)
+		if err != nil {
+			result = multierror.Append(result, fmt.Errorf("failed to eval labels for metric '%s' with err: %s", name, err))
+			continue
+		}
+
+		// TODO: investigate either pooling these, or keeping a set around that has only its field's values updated.
+		// we could keep a map[metric name]value, iterate over the it updating only the fields in each value
+		values = append(values, adapter.Value{
+			Name:   name,
+			Kind:   md.metricKind,
+			Labels: labels,
+			// TODO: extract standard timestamp attributes for start/end once we det'm what they are
+			StartTime:   time.Now(),
+			EndTime:     time.Now(),
+			MetricValue: metricValue,
+		})
+	}
+
+	if err := w.aspect.Record(values); err != nil {
+		result = multierror.Append(result, fmt.Errorf("failed to record all values with err: %s", err))
+	}
+
+	// We may accumulate errors while still being able to record some metrics; if we have any values at all we'll
+	// return an OK alongside our errors (since presumably we're able to record at least a few).
+	var out *Output
+	if len(values) > 0 {
+		out = &Output{Code: code.Code_OK}
```

Comment:
nit: should we set a default `*Output` with  `Code_OK` and then have this check actually override with something like `Code_INVALID_ARGUMENT` for when `len(values) == 0`?